Compilation failed on ppc64

Hi All, I am trying to compile Julia 0.4.5 on Power8, Redhat 7.2/64-bit. My "Make.user" file:

    prefix=${SOFT}/julia  
[julia.ppc64.log.zip](https://github.com/JuliaLang/julia/files/234398/julia.ppc64.log.zip)

                  # <-- custom destination folder, SOFT is defined in .bashrc
    include $(JULIAHOME)/Make.powerpc

The failure happens while "inference.jl" is compiled (the complete output file is attached):

```
inference.jl
julia: /home/albert/tmp/julia-0.4.5/deps/llvm-3.6.0/lib/CodeGen/MachineLICM.cpp:960: bool {anonymous}::MachineLICM::IsLoopInvariantInst(llvm::MachineInstr&): Assertion `MRI->getVRegDef(Reg) && "Machine instr not mapped for this vreg?!"' failed.

signal (6): Aborted
/bin/sh: line 1: 102713 Aborted                 (core dumped) /home/albert/tmp/julia-0.4.5/usr/bin/julia -C native --output-ji /home/albert/tmp/julia-0.4.5/usr/lib/julia/inference0.ji -f coreimg.jl
make[1]: *** [/home/albert/tmp/julia-0.4.5/usr/lib/julia/inference0.ji] Error 134
make: *** [julia-inference] Error 2
```

Hi All, I tried with the source from GitHub, the problem happens at a bit earlier stage than before (note, llvm-3.7.1 is used for now):

/home/albert/tmp/julia/src/ccall.cpp:127:4: warning: #warning is a GCC extension [enabled by default]
 #  warning "ccall is defaulting to llvm ABI, since no platform ABI has been defined for this CPU/OS combination"
    ^
/home/albert/tmp/julia/src/ccall.cpp:127:4: warning: #warning "ccall is defaulting to llvm ABI, since no platform ABI has been defined for this CPU/OS combination" [-Wcpp]
    CC src/disasm.o
    CC src/debuginfo.o
    CC src/llvm-simdloop.o
    CC src/llvm-gcroot.o
    LINK usr/lib/libjulia.so
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Warning: File `../src/julia_version.h' has modification time 84 s in the future
    CC ui/repl.o
    LINK usr/bin/julia
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
Creating usr/etc/julia/juliarc.jl
Copying in usr/share/man/man1/julia.1
    PERL base/pcre_h.jl
    PERL base/errno_h.jl
    PERL base/build_h.jl.phony
    PERL base/fenv_constants.jl
    PERL base/file_constants.jl
    PERL base/uv_constants.jl
    PERL base/version_git.jl.phony
Copying in usr/share/doc/julia/examples
    JULIA usr/lib/julia/inference.ji
julia: /home/albert/tmp/julia/deps/srccache/llvm-3.7.1/include/llvm/Support/Casting.h:237: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = llvm::MCSymbolELF; Y = llvm::MCSymbol; typename llvm::cast_retty<X, Y*>::ret_type = llvm::MCSymbolELF*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
/bin/sh: line 1: 196533 Aborted                 (core dumped) /home/albert/tmp/julia/usr/bin/julia -C native --output-ji /home/albert/tmp/julia/usr/lib/julia/inference.ji -f coreimg.jl
make[1]: *** [/home/albert/tmp/julia/usr/lib/julia/inference.ji] Error 134
make: *** [julia-inference] Error 2

--- a/contrib/add_license_to_files.jl
+++ b/contrib/add_license_to_files.jl
@@ -38,6 +38,7 @@ const skipfiles = [
     "../base/linalg/givens.jl",
     #
     "../src/abi_llvm.cpp",
+    "../src/abi_ppc64le.cpp",
     "../src/abi_win32.cpp",
     "../src/abi_win64.cpp",
     "../src/abi_x86.cpp",

--- a/deps/llvm-3.8.0_ppc64_SUBFC8.patch
+++ b/deps/llvm-3.8.0_ppc64_SUBFC8.patch
@@ -1,0 +1,31 @@
+diff --git a/lib/Target/PowerPC/PPCInstr64Bit.td b/lib/Target/PowerPC/PPCInstr64Bit.td
+index 52af293..e7eb8a1 100644
+--- a/lib/Target/PowerPC/PPCInstr64Bit.td
++++ b/lib/Target/PowerPC/PPCInstr64Bit.td
+@@ -514,11 +514,11 @@ let Defs = [CARRY] in {
+ def SUBFIC8: DForm_2< 8, (outs g8rc:$rD), (ins g8rc:$rA, s16imm64:$imm),
+                      "subfic $rD, $rA, $imm", IIC_IntGeneral,
+                      [(set i64:$rD, (subc imm64SExt16:$imm, i64:$rA))]>;
+-defm SUBFC8 : XOForm_1r<31, 8, 0, (outs g8rc:$rT), (ins g8rc:$rA, g8rc:$rB),
++}
++defm SUBFC8 : XOForm_1rc<31, 8, 0, (outs g8rc:$rT), (ins g8rc:$rA, g8rc:$rB),
+                         "subfc", "$rT, $rA, $rB", IIC_IntGeneral,
+                         [(set i64:$rT, (subc i64:$rB, i64:$rA))]>,
+                         PPC970_DGroup_Cracked;
+-}
+ defm SUBF8 : XOForm_1r<31, 40, 0, (outs g8rc:$rT), (ins g8rc:$rA, g8rc:$rB),
+                        "subf", "$rT, $rA, $rB", IIC_IntGeneral,
+                        [(set i64:$rT, (sub i64:$rB, i64:$rA))]>;
+diff --git a/lib/Target/PowerPC/PPCInstrInfo.cpp b/lib/Target/PowerPC/PPCInstrInfo.cpp
+index 313ab88..8dca286 100644
+--- a/lib/Target/PowerPC/PPCInstrInfo.cpp
++++ b/lib/Target/PowerPC/PPCInstrInfo.cpp
+@@ -1795,6 +1795,7 @@ bool PPCInstrInfo::optimizeCompareInstr(MachineInstr *CmpInstr,
+           MI->addOperand(*MI->getParent()->getParent(),
+                          MachineOperand::CreateReg(*ImpUses, false, true));
+   }
++  assert(MI->definesRegister(PPC::CR0));
+
+   // Modify the condition code of operands in OperandsToUpdate.
+   // Since we have SUB(r1, r2) and CMP(r2, r1), the condition code needs to
+

--- a/deps/llvm.mk
+++ b/deps/llvm.mk
@@ -420,6 +420,7 @@ $(LLVM_SRC_DIR)/llvm-D17326_unpack_load.patch-applied: $(LLVM_SRC_DIR)/llvm-D142
 $(eval $(call LLVM_PATCH,llvm-D17712))
 $(eval $(call LLVM_PATCH,llvm-PR26180))
 $(eval $(call LLVM_PATCH,llvm-PR27046))
+$(eval $(call LLVM_PATCH,llvm-3.8.0_ppc64_SUBFC8))
 endif # LLVM_VER
 
 ifeq ($(LLVM_VER),3.7.1)

--- a/src/abi_ppc64le.cpp
+++ b/src/abi_ppc64le.cpp
@@ -1,0 +1,90 @@
+//===-- abi_ppc64le.cpp - Power v2 ABI description ---------------------*- C++ -*-===//
+//
+//                         LDC – the LLVM D compiler
+//
+// This file is distributed under the BSD-style LDC license:
+//
+// Copyright (c) 2007-2012 LDC Team.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright notice,
+//       this list of conditions and the following disclaimer in the documentation
+//       and/or other materials provided with the distribution.
+//     * Neither the name of the LDC Team nor the names of its contributors may be
+//       used to endorse or promote products derived from this software without
+//       specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//===----------------------------------------------------------------------===//
+//
+// The ABI implementation used for 64 bit little-endian PowerPC targets.
+//
+// The PowerOpen 64bit ELF v2 ABI can be found here:
+// https://members.openpowerfoundation.org/document/dl/576
+//===----------------------------------------------------------------------===//
+
+
+typedef bool AbiState;
+AbiState default_abi_state = 0;
+
+static bool isHFA(jl_datatype_t *ty)
+{
+    size_t i, l = ty->nfields;
+    jl_value_t *fld0 = jl_field_type(ty, 0);
+    if (l > 8)
+        return false;
+    if (fld0 != (jl_value_t*)jl_float64_type && fld0 != (jl_value_t*)jl_float32_type)
+        return false;
+    for (i = 1; i < l; i++) {
+        jl_value_t *fld = jl_field_type(ty, i);
+        if (fld != fld0)
+            return false;
+    }
+    return true;
+}
+
+bool use_sret(AbiState *state, jl_value_t *ty)
+{
+    // Assume jl_is_datatype(ty) && !jl_is_abstracttype(ty)
+    jl_datatype_t *dt = (jl_datatype_t*)ty;
+    if (dt->size > 16 && !isHFA(dt))
+        return true;
+    return false;
+}
+
+void needPassByRef(AbiState *state, jl_value_t *ty, bool *byRef, bool *inReg)
+{
+    *byRef = false;
+}
+
+Type *preferred_llvm_type(jl_value_t *ty, bool isret)
+{
+    // Arguments are either scalar or passed by value
+    if (!jl_is_datatype(ty) || jl_is_abstracttype(ty))
+        return NULL;
+    jl_datatype_t *dt = (jl_datatype_t*)ty;
+    // rewrite integer sized (non-HFA) struct to the corresponding integer
+    if (!dt->nfields || (isret ? use_sret(NULL, ty) : false) || isHFA(dt))
+        return NULL;
+    return Type::getIntNTy(jl_LLVMContext, dt->size * 8);
+}
+
+bool need_private_copy(jl_value_t *ty, bool byRef)
+{
+    return false;
+}

--- a/src/abi_ppc64le.cpp
+++ b/src/abi_ppc64le.cpp
@@ -42,46 +42,67 @@
 typedef bool AbiState;
 AbiState default_abi_state = 0;
 
-static bool isHFA(jl_datatype_t *ty)
+// count the homogeneous floating agregate size (saturating at max count of 8)
+static unsigned isHFA(jl_datatype_t *ty, jl_datatype_t **ty0)
 {
     size_t i, l = ty->nfields;
-    jl_value_t *fld0 = jl_field_type(ty, 0);
-    if (l > 8)
-        return false;
-    if (fld0 != (jl_value_t*)jl_float64_type && fld0 != (jl_value_t*)jl_float32_type)
-        return false;
-    for (i = 1; i < l; i++) {
-        jl_value_t *fld = jl_field_type(ty, i);
-        if (fld != fld0)
-            return false;
+    if (l == 0) {
+        if (*ty0 == NULL) {
+            if (ty == jl_float64_type || ty == jl_float32_type)
+                *ty0 = ty;
+        }
+        return ty == *ty0 ? 1 : 9;
     }
-    return true;
+    int n = 0;
+    for (i = 0; i < l; i++) {
+        jl_value_t *fld = jl_field_type(ty, i);
+        if (!jl_is_datatype(fld))
+            return 9;
+        n += isHFA((jl_datatype_t*)fld, ty0);
+        if (n > 8)
+            return 9;
+    }
+    return n;
 }
 
 bool use_sret(AbiState *state, jl_value_t *ty)
 {
     // Assume jl_is_datatype(ty) && !jl_is_abstracttype(ty)
     jl_datatype_t *dt = (jl_datatype_t*)ty;
-    if (dt->size > 16 && !isHFA(dt))
+    jl_datatype_t *ty0 = NULL;
+    if (dt->size > 16 && isHFA(dt, &ty0) > 8)
         return true;
     return false;
 }
 
 void needPassByRef(AbiState *state, jl_value_t *ty, bool *byRef, bool *inReg)
 {
-    *byRef = false;
+    if (!jl_is_datatype(ty) || jl_is_abstracttype(ty) || jl_is_cpointer_type(ty) || jl_is_array_type(ty))
+        return;
+    size_t size = jl_datatype_size(ty);
+    if (size > 64)
+        *byRef = true;
 }
 
 Type *preferred_llvm_type(jl_value_t *ty, bool isret)
 {
     // Arguments are either scalar or passed by value
-    if (!jl_is_datatype(ty) || jl_is_abstracttype(ty))
+    if (!jl_is_datatype(ty) || jl_is_abstracttype(ty) || jl_is_cpointer_type(ty) || jl_is_array_type(ty))
         return NULL;
     jl_datatype_t *dt = (jl_datatype_t*)ty;
-    // rewrite integer sized (non-HFA) struct to the corresponding integer
-    if (!dt->nfields || (isret ? use_sret(NULL, ty) : false) || isHFA(dt))
+    size_t size = dt->size;
+    // don't need to change bitstypes
+    if (!dt->nfields)
         return NULL;
-    return Type::getIntNTy(jl_LLVMContext, dt->size * 8);
+    // legalize this into [n x f32/f64]
+    jl_datatype_t *ty0 = NULL;
+    int hfa = isHFA(dt, &ty0);
+    if (hfa <= 8)
+        return ArrayType::get(ty0 == jl_float32_type ? T_float32 : T_float64, hfa);
+    // rewrite integer-sized (non-HFA) struct to an array of i64
+    if (size > 8)
+        return ArrayType::get(T_int64, (size + 7) / 8);
+    return Type::getIntNTy(jl_LLVMContext, size * 8);
 }
 
 bool need_private_copy(jl_value_t *ty, bool byRef)

--- a/src/abi_win32.cpp
+++ b/src/abi_win32.cpp
@@ -67,7 +67,7 @@ Type *preferred_llvm_type(jl_value_t *ty, bool isret)
         return NULL;
     jl_datatype_t *dt = (jl_datatype_t*)ty;
     // rewrite integer sized (non-sret) struct to the corresponding integer
-    if (!dt->nfields || use_sret(NULL, ty))
+    if (!dt->nfields)
         return NULL;
     return Type::getIntNTy(jl_LLVMContext, dt->size * 8);
 }

--- a/src/ccall.cpp
+++ b/src/ccall.cpp
@@ -122,7 +122,9 @@ static Value *runtime_sym_lookup(PointerType *funcptype, const char *f_lib, cons
 #elif defined _CPU_ARM_
 #  include "abi_arm.cpp"
 #elif defined _CPU_AARCH64_
-#    include "abi_aarch64.cpp"
+#  include "abi_aarch64.cpp"
+#elif defined _CPU_PPC64_
+#  include "abi_ppc64le.cpp"
 #else
 #  warning "ccall is defaulting to llvm ABI, since no platform ABI has been defined for this CPU/OS combination"
 #  include "abi_llvm.cpp"

--- a/src/ccall.cpp
+++ b/src/ccall.cpp
@@ -876,7 +876,6 @@ static std::string generate_func_sig(
     size_t nargt = jl_svec_len(tt);
     assert(rt && !jl_is_abstract_ref_type(rt));
 
-    AttrBuilder retattrs;
     std::vector<AttrBuilder> paramattrs;
     AbiState abi = default_abi_state;
     sret = 0;
@@ -885,7 +884,7 @@ static std::string generate_func_sig(
         *prt = *lrt = T_void;
     }
     else {
-        *prt = preferred_llvm_type(rt, true);
+        *prt = sret ? NULL : preferred_llvm_type(rt, true);
         if (*prt == NULL)
             *prt = *lrt;
 
@@ -895,6 +894,7 @@ static std::string generate_func_sig(
 #if !defined(_OS_WINDOWS_) || defined(LLVM35) // llvm used to use the old mingw ABI, skipping this marking works around that difference
             paramattrs[0].addAttribute(Attribute::StructRet);
 #endif
+            paramattrs[0].addAttribute(Attribute::NoAlias);
             fargt_sig.push_back(PointerType::get(*prt, 0));
             sret = 1;
         }
@@ -953,14 +953,8 @@ static std::string generate_func_sig(
             needPassByRef(&abi, tti, &byRef, &inReg);
         }
 
-        Type *pat = preferred_llvm_type(tti, false);
-        if (pat != NULL) {
-            assert(!byRef); // it is an error for an ABI to specify a preferred type for a pointer arg
-        }
-        else if (byRef) {
-            pat = PointerType::get(t, 0);
-        }
-        else {
+        Type *pat = byRef ? PointerType::get(t, 0) : preferred_llvm_type(tti, false);
+        if (pat == NULL) {
             pat = t;
         }
 
@@ -992,10 +986,6 @@ static std::string generate_func_sig(
             }
             i++;
         } while (current_isVa && i < nargs); // if is this is the vararg, loop to the end
-    }
-
-    if (retattrs.hasAttributes()) {
-        attributes = AttributeSet::get(jl_LLVMContext, AttributeSet::ReturnIndex, retattrs);
     }
 
     for (i = 0; i < nargs + sret; ++i) {

--- a/src/ccalltest.c
+++ b/src/ccalltest.c
@@ -196,6 +196,125 @@ typedef struct {
     char z;
 } struct_big;
 
+typedef struct {
+    int64_t r1;
+    int64_t r2;
+    int64_t r3;
+    int64_t r4;
+    int64_t r5;
+    int64_t r6;
+    int64_t r7;
+    int64_t r8;
+} struct_huge1a;
+
+typedef struct {
+    int64_t r1;
+    int64_t r2;
+    int64_t r3;
+    int64_t r4;
+    int64_t r5;
+    int64_t r6;
+    int64_t r7;
+    int64_t r8;
+    int64_t r9;
+} struct_huge1b;
+
+typedef struct {
+    double f1;
+    double f2;
+    double f3;
+    double f4;
+    double f5;
+    double f6;
+    double f7;
+    double f8;
+} struct_huge2a;
+
+typedef struct {
+    double f1;
+    double f2;
+    double f3;
+    double f4;
+    double f5;
+    double f6;
+    double f7;
+    double f8;
+    double f9;
+} struct_huge2b;
+
+typedef struct {
+    complex float f12;
+    complex float f34;
+    complex float f56;
+    float f7;
+    float f8;
+} struct_huge3a;
+
+typedef struct {
+    complex float r1;
+    complex float r2;
+    complex float r3;
+    complex float r4;
+    complex float r5;
+    complex float r6;
+    complex float r7;
+    float r8a;
+    float r8b;
+} struct_huge3b;
+
+typedef struct {
+    complex float r1;
+    complex float r2;
+    complex float r3;
+    complex float r4;
+    complex float r5;
+    complex float r6;
+    complex float r7;
+    float r8a;
+    float r8b;
+    float r9;
+} struct_huge3c;
+
+typedef struct {
+    complex double r12;
+    complex double r34;
+    complex float r5;
+    complex double r67;
+    double r8;
+} struct_huge4a;
+
+typedef struct {
+    complex double r12;
+    complex double r34;
+    complex float r5;
+    complex double r67;
+    complex double r89;
+} struct_huge4b;
+
+typedef struct {
+    complex int r1;
+    complex int r2;
+    complex int r3;
+    complex int r4;
+    complex int r5;
+    complex int r6;
+    complex int r7;
+    complex int r8;
+} struct_huge5a;
+
+typedef struct {
+    complex int r1;
+    complex int r2;
+    complex int r3;
+    complex int r4;
+    complex int r5;
+    complex int r6;
+    complex int r7;
+    complex int r8;
+    complex int r9;
+} struct_huge5b;
+
+
 JL_DLLEXPORT struct1 test_1(struct1 a, float b) {
     //Unpack a "small" struct { float, double }
     if (verbose) fprintf(stderr,"%g + %g i & %g\n", a.x, a.y, b);
@@ -366,6 +485,25 @@ JL_DLLEXPORT struct_big test_big(struct_big a) {
     a.z -= 'A';
     return a;
 }
+
+#define test_huge(suffix, reg) \
+JL_DLLEXPORT struct_huge##suffix test_huge##suffix(char a, struct_huge##suffix b, char c) { \
+    if (verbose) fprintf(stderr,"%c-%c\n", a, c); \
+    b.reg *= 39; \
+    return b; \
+}
+
+test_huge(1a, r1);
+test_huge(1b, r1);
+test_huge(2a, f1);
+test_huge(2b, f1);
+test_huge(3a, f12);
+test_huge(3b, r1);
+test_huge(3c, r1);
+test_huge(4a, r12);
+test_huge(4b, r12);
+test_huge(5a, r1);
+test_huge(5b, r1);
 
 JL_DLLEXPORT int get_c_int(void)
 {

--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -4126,8 +4126,10 @@ static std::unique_ptr<Module> emit_function(jl_lambda_info_t *lam, jl_llvm_func
         f = Function::Create(FunctionType::get(rt, fsig, false),
                              GlobalVariable::ExternalLinkage,
                              funcName.str(), M);
-        if (ctx.sret)
+        if (ctx.sret) {
             f->addAttribute(1, Attribute::StructRet);
+            f->addAttribute(1, Attribute::NoAlias);
+        }
         addComdat(f);
 #ifdef LLVM37
         f->addFnAttr("no-frame-pointer-elim", "true");

--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -5743,13 +5743,13 @@ extern "C" void jl_init_codegen(void)
         .setMCJITMemoryManager(std::unique_ptr<RTDyldMemoryManager>{new SectionMemoryManager()})
 #endif
         .setTargetOptions(options)
-#if (defined(_OS_LINUX_) && defined(_CPU_X86_64_)) || defined(CODEGEN_TLS)
+#if (defined(_OS_LINUX_) && defined(_CPU_X86_64_))
         .setRelocationModel(Reloc::PIC_)
 #elif !defined(LLVM39)
         .setRelocationModel(Reloc::Default)
 #endif
-#ifdef CODEGEN_TLS
-        .setCodeModel(CodeModel::Small)
+#ifdef _P64
+        .setCodeModel(CodeModel::Large)
 #else
         .setCodeModel(CodeModel::JITDefault)
 #endif

--- a/src/debuginfo.cpp
+++ b/src/debuginfo.cpp
@@ -1515,8 +1515,10 @@ static DW_EH_PE parseCIE(const uint8_t *Addr, const uint8_t *End)
     // Assume there's no EH Data field, which exist when the augmentation
     // string has "eh" in it.
     p += augmentation_len + 1;
-    // Code Alignment Factor (1 byte) should always be 1
-    assert(*p == 1);
+    // Code Alignment Factor (1 byte)
+    // should always be 1 on x86, 4 on PPC, etc.
+    // (used for DW_CFA_advance_loc / not used here)
+    //assert(*p == 1);
     p++;
     // Data Alignment Factor (LEB128)
     assert(cie_end >= p);

--- a/src/options.h
+++ b/src/options.h
@@ -113,9 +113,9 @@
 #  define MEMDEBUG
 #  define KEEP_BODIES
 #  endif
-// Memory sanitizer also needs thread-local storage
+// Memory sanitizer needs TLS, which llvm only supports for the small memory model
 #  if __has_feature(memory_sanitizer)
-#  define CODEGEN_TLS
+   // todo: fix the llvm MemoryManager to work with small memory model
 #  endif
 #endif
 

--- a/src/stackwalk.c
+++ b/src/stackwalk.c
@@ -280,7 +280,7 @@ static int jl_unw_step(bt_cursor_t *cursor, uintptr_t *ip, uintptr_t *sp)
 #endif
 }
 
-#elif defined(_CPU_ARM_) || defined(_CPU_PPC64_)
+#elif defined(_CPU_ARM_)
 // platforms on which libunwind may be broken
 
 static int jl_unw_init(bt_cursor_t *cursor, bt_context_t *context) { return 0; }

--- a/test/ccall.jl
+++ b/test/ccall.jl
@@ -397,6 +397,83 @@ let
     @test x.z == a.z - Int('A')
 end
 
+const Struct_huge1a = NTuple{8, Int64}
+const Struct_huge1b = NTuple{9, Int64}
+const Struct_huge2a = NTuple{8, Cdouble}
+const Struct_huge2b = NTuple{9, Cdouble}
+type Struct_huge3a
+    cf::NTuple{3, Complex{Cfloat}}
+    f7::Cfloat
+    f8::Cfloat
+end
+type Struct_huge3b
+    cf::NTuple{7, Complex{Cfloat}}
+    r8a::Cfloat
+    r8b::Cfloat
+end
+type Struct_huge3c
+    cf::NTuple{7, Complex{Cfloat}}
+    r8a::Cfloat
+    r8b::Cfloat
+    r9::Cfloat
+end
+type Struct_huge4a
+    r12::Complex{Cdouble}
+    r34::Complex{Cdouble}
+    r5::Complex{Cfloat}
+    r67::Complex{Cdouble}
+    r8::Cdouble
+end
+type Struct_huge4b
+    r12::Complex{Cdouble}
+    r34::Complex{Cdouble}
+    r5::Complex{Cfloat}
+    r67::Complex{Cdouble}
+    r89::Complex{Cdouble}
+end
+const Struct_huge5a = NTuple{8, Complex{Cint}}
+const Struct_huge5b = NTuple{9, Complex{Cint}}
+
+function verify_huge(init, a, b)
+    @test typeof(init) === typeof(a) === typeof(b)
+    verbose && @show (a, b)
+    # make sure a was unmodified
+    for i = 1:nfields(a)
+        @test getfield(init, i) === getfield(a, i)
+    end
+    # make sure b was modifed as expected
+    a1, b1 = getfield(a, 1), getfield(b, 1)
+    if isa(a1, Tuple)
+        @test oftype(a1[1], a1[1] * 39) === b1[1]
+        @test a1[2:end] === b1[2:end]
+    else
+        @test oftype(a1, a1 * 39) === b1
+    end
+    for i = 2:nfields(a)
+        @test getfield(a, i) === getfield(b, i)
+    end
+end
+macro test_huge(i, b, init)
+    f = QuoteNode(Symbol("test_huge", i, b))
+    ty = Symbol("Struct_huge", i, b)
+    return quote
+        let a = $ty($(esc(init))...), f
+            f(b) = ccall(($f, libccalltest), $ty, (Cchar, $ty, Cchar), '0' + $i, a, $b)
+            verify_huge($ty($(esc(init))...), a, f(a))
+        end
+    end
+end
+@test_huge 1 'a' ((1, 2, 3, 4, 5, 6, 7, 8),)
+@test_huge 1 'b' ((1, 2, 3, 4, 5, 6, 7, 8, 9),)
+@test_huge 2 'a' ((1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0),)
+@test_huge 2 'b' ((1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0),)
+@test_huge 3 'a' ((1.0 + 2.0im, 3.0 + 4.0im, 5.0 + 6.0im), 7.0, 8.0)
+@test_huge 3 'b' ((1.0 + 2.0im, 3.0 + 4.0im, 5.0 + 6.0im, 7.0 + 8.0im, 9.0 + 10.0im, 11.0 + 12.0im, 13.0 + 14.0im), 7.0, 8.0)
+@test_huge 3 'c' ((1.0 + 2.0im, 3.0 + 4.0im, 5.0 + 6.0im, 7.0 + 8.0im, 9.0 + 10.0im, 11.0 + 12.0im, 13.0 + 14.0im), 7.0, 8.0, 9.0)
+@test_huge 4 'a' (1.0 + 2.0im, 3.0 + 4.0im, 5.0f0 + 6.0f0im, 7.0 + 8.0im, 9.0)
+@test_huge 4 'b' (1.0 + 2.0im, 3.0 + 4.0im, 5.0f0 + 6.0f0im, 7.0 + 8.0im, 9.0 + 10.0im)
+@test_huge 5 'a' ((1 + 2im, 3 + 4im, 5 + 6im, 7 + 8im, 9 + 10im, 11 + 12im, 13 + 14im, 15 + 16im),)
+@test_huge 5 'b' ((1 + 2im, 3 + 4im, 5 + 6im, 7 + 8im, 9 + 10im, 11 + 12im, 13 + 14im, 15 + 16im, 17 + 17im),)
 
 ## cfunction roundtrip
 

